[css-anchor-position-1] Last-successful position fallback doesn't apply on anchoreum.com

tuankiet65 opened the pull request:
#### 9a79c5246dd6cb20a321af81ac1e2f2eb432a705
<pre>
[css-anchor-position-1] Last-successful position fallback doesn&apos;t apply on anchoreum.com
<a href="https://rdar.apple.com/158365856">rdar://158365856</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297438">https://bugs.webkit.org/show_bug.cgi?id=297438</a>

Reviewed by Antti Koivisto.

If a PositionTryFallback is a position-area option, the position-area is
stored in a RefPtr&lt;const StyleProperties&gt;. The current operator== for
PositionTryFallback is default, so it&apos;d only check whether the StyleProperties
pointers are the same. Fix it to look inside the StyleProperties and check
if the CSSValue of position-area are equal.

Also improve operator&lt;&lt; of PositionTryFallback to print out more debug
information.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-position-area-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-position-area.html: Added.
* Source/WebCore/rendering/style/PositionTryFallback.cpp:
(WebCore::Style::operator==):
(WebCore::Style::operator&lt;&lt;):
* Source/WebCore/rendering/style/PositionTryFallback.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
- Use ImmutableStyleProperties::createStyleDeduplicating to deduplicate the
  returning ImmutableStyleProperties, if an identical one was created before.

Canonical link: <a href="https://commits.webkit.org/299367@main">https://commits.webkit.org/299367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3edec53611c2c60cdcadef2750032b25305e854

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29089 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124942 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70817 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/17d6a75e-9f81-4513-b7f1-e4ae1b82c529) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120635 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47020 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90131 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59642 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ec8fd6c8-c320-49f3-b71b-00a66a10d534) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121710 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31179 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106470 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70637 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2e4879f7-e617-481f-a01c-cfb7d0b5d8a8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30239 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24583 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68601 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100622 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127996 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34469 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98779 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46028 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102690 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98561 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43999 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22005 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42187 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45534 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51212 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44998 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48344 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46684 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jesxilin commented on Source/WebCore/rendering/style/PositionTryFallback.cpp:
```diff
@@ -26,32 +26,122 @@
+const std::optional<ScopedName>& PositionTryFallback::positionTryRuleName() const
+{
+    ASSERT(m_type == Type::RuleAndTactic);
+
```
Nit suggestion: I would remove the extra newline and on other instances.

tuankiet65 replied:
Maybe it's just me, but my eyes couldn't read many lines of code clumped together, so I like to sprinkle newlines around to make stuff more readable 😅 But I'll look for some newlines to get rid of

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -1528,11 +1528,14 @@ std::unique_ptr<RenderStyle> TreeResolver::generatePositionOption(const Position
-            return fallback.positionAreaProperties;
+
+            auto property = CSSProperty { CSSPropertyPositionArea, Ref { *fallback.positionArea } };
+            return ImmutableStyleProperties::create(std::span { &property, 1 }, HTMLStandardMode);
```
Why are we doing this? We were caching ImmutableStyleProperties and now we aren't?

tuankiet65 committed:
[css-anchor-position-1] Last-successful position fallback doesn't apply on anchoreum.com
rdar://158365856
https://bugs.webkit.org/show_bug.cgi?id=297438

Reviewed by Antti Koivisto.

If a PositionTryFallback is a position-area option, the position-area is
stored in a RefPtr<const StyleProperties>. The current operator== for
PositionTryFallback is default, so it'd only check whether the StyleProperties
pointers are the same. Fix it to look inside the StyleProperties and check
if the CSSValue of position-area are equal.

Also improve operator<< of PositionTryFallback to print out more debug
information.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-position-area-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-position-area.html: Added.
* Source/WebCore/rendering/style/PositionTryFallback.cpp:
(WebCore::Style::operator==):
(WebCore::Style::operator<<):
* Source/WebCore/rendering/style/PositionTryFallback.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
- Use ImmutableStyleProperties::createStyleDeduplicating to deduplicate the
  returning ImmutableStyleProperties, if an identical one was created before.

Canonical link: https://commits.webkit.org/299367@main
